docs/dev/data-identity.rst

```diff
@@ -0,0 +1,250 @@
+
+      * ``enum_identity`` C++ ``enum``
+
+      * ``struct_identity`` C++ ``class`` or ``structure``
+
+        * ``global_identity`` holds, as a quasiobject, handles for all of the known Dwarf Fortress program-scope static objects as if they were fields of an object called ``global``
+
+        * ``union_identity`` C++ ``union``
+
+        * ``other_vectors_identity`` special-case identity for the categorized subvectors of objects that appears in many of Dwarf Fortress's "handler" classes
+
+        * ``virtual_identity`` polymorphic C++ ``class`` or ``structure`` having a virtual table to handle virtual dispatch
+
+      * ``stl_string_identity`` ``std::string``
+
+      * ``xlsx_file_handle_identity`` special case
+
+      * ``xlsx_sheet_handle_identity`` special case
+
+    * ``container_identity`` "containers" generally. note that all container types are homogeneous (that is, the elements of the container must all be of the same type). abstract base class
+
+      * ``bit_container_identity`` for containers that contain bools stored one element per *bit* (rather than per byte)
+
+        * ``bit_array_identity`` Dwarf Fortress's ``BitArray`` type
+
+        * ``stl_bit_vector_identity`` ``std::vector<bool>``
+
+      * ``buffer_container_identity`` C++ static arrays and raw C++ pointers acting as arrays of unspecified bound
+
+      * ``enum_list_attr_identity`` (template) metaobject with metadata about a C++ enumeration; may also include additional metadata
+
+      * ``ptr_container_identity`` containers that contain pointers
+
+        *  ``stl_ptr_container_identity`` containers that are of the form ``std::vector<T*>`` for some ``T``
+
+      * ``ro_stl_container_identity`` (template) "read only containers"
+
+        * ``ro_stl_assoc_container_identity`` (template) ``std::map<KT,T>`` and ``std::unordered_map<KT,T>``
+
+      * ``stl_container_identity`` (template) ``std::vector<T>`` where ``T`` is *not* a pointer (and not ``bool``)
+
+    * ``opaque_identity`` opaque wrapper around any type, provides no functionality
+
+    * ``stl_string_identity`` ``std::string``
+
+  * ``function_identity_base`` abstract base class for ``function_identity``
+
+    * ``function_identity`` (template) wrapper around a C++ function that can be invoked from Lua
+
+  * ``primitive_identity`` wrapper around a primitive type. primitive types are fixed-length objects with no internal structure
+
+    * ``bool_identity`` ``bool``
+
+    * ``number_identity_base`` abstract base for numeric types
+
+      * ``float_identity_base`` abstract base for floating point types
+
+        * ``float_identity`` (template) ``double`` and ``float``
+
+      * ``integer_identity_base`` abstract base for integral types
+
+        * ``integer_identity`` (template) ``int8_t``, ``int16_t``, ``int32_t``, ``size_t``, etc. lots of these
+
+    * ``pointer_identity`` any arbitrary C++ pointer (other than ``char*``)
+
+    * ``ptr_string_identity`` C-style (``char *``) string
```

Comment:
```suggestion
* ``type_identity``

  * ``constructed_identity`` anything with an internal structure (not primitive)
  
    * ``compound_identity`` anything with fields

      * ``bitfield_identity`` a structure defined with fields at bit rather than byte boundaries
      * ``enum_identity`` C++ ``enum``
      * ``struct_identity`` C++ ``class`` or ``structure``

        * ``global_identity`` holds, as a quasiobject, handles for all of the known Dwarf Fortress program-scope static objects as if they were fields of an object called ``global``
        * ``union_identity`` C++ ``union``
        * ``other_vectors_identity`` special-case identity for the categorized subvectors of objects that appears in many of Dwarf Fortress's "handler" classes
        * ``virtual_identity`` polymorphic C++ ``class`` or ``structure`` having a virtual table to handle virtual dispatch

      * ``stl_string_identity`` ``std::string``
      * ``xlsx_file_handle_identity`` special case
      * ``xlsx_sheet_handle_identity`` special case

    * ``container_identity`` "containers" generally. note that all container types are homogeneous (that is, the elements of the container must all be of the same type). abstract base class

      * ``bit_container_identity`` for containers that contain bools stored one element per *bit* (rather than per byte)

        * ``bit_array_identity`` Dwarf Fortress's ``BitArray`` type
        * ``stl_bit_vector_identity`` ``std::vector<bool>``

      * ``buffer_container_identity`` C++ static arrays and raw C++ pointers acting as arrays of unspecified bound
      * ``enum_list_attr_identity`` (template) metaobject with metadata about a C++ enumeration; may also include additional metadata

      * ``ptr_container_identity`` containers that contain pointers

        *  ``stl_ptr_container_identity`` containers that are of the form ``std::vector<T*>`` for some ``T``

      * ``ro_stl_container_identity`` (template) "read only containers"

        * ``ro_stl_assoc_container_identity`` (template) ``std::map<KT,T>`` and ``std::unordered_map<KT,T>``

      * ``stl_container_identity`` (template) ``std::vector<T>`` where ``T`` is *not* a pointer (and not ``bool``)

    * ``opaque_identity`` opaque wrapper around any type, provides no functionality
    * ``stl_string_identity`` ``std::string``

  * ``function_identity_base`` abstract base class for ``function_identity``

    * ``function_identity`` (template) wrapper around a C++ function that can be invoked from Lua

  * ``primitive_identity`` wrapper around a primitive type. primitive types are fixed-length objects with no internal structure

    * ``bool_identity`` ``bool``
    * ``number_identity_base`` abstract base for numeric types

      * ``float_identity_base`` abstract base for floating point types

        * ``float_identity`` (template) ``double`` and ``float``

      * ``integer_identity_base`` abstract base for integral types

        * ``integer_identity`` (template) ``int8_t``, ``int16_t``, ``int32_t``, ``size_t``, etc. lots of these

    * ``pointer_identity`` any arbitrary C++ pointer (other than ``char*``)
    * ``ptr_string_identity`` C-style (``char *``) string
```

Reply:
I'm not sure this is actually any better, but you could shave off a few newlines if you wanted to